add v0 tech spec (#197)
chanzuckerberg/napari-hub: docs/v0-tech-spec.md

* Ability for users to search for a plugin by words (i.e. global search bar with plugins as results)
* Ability to browse existing plugins organized by type(s)
* Technical requirements:
  * Page and plugin data should load in a reasonable amount of time.
  * Search/list functionality should be able to support 100s of plugins (closer to the lower end)  for the first version. 
  * Search inputs should have form validation.
  * Search form fields:
    * Global search [input/textbox] - Spreadsheet with fields and info defined. We will use all items from the “Search” 
      column. More details for this feature defined under Features: Global Search Bar.
    * Filter [checkboxes] - Spreadsheet with fields and info defined. We will use all items from the “Filter” column. 
      More details for this feature defined under Features: Filter/Sort
    * Sort [dropdown/radio buttons] - Spreadsheet with fields and info defined. We will use all items from the “Sort” 
      column. More details for this feature defined under Features: Filter/Sort
  * Plugin list data to display
    * Fields and info specification on Github. We will use all items from the “List view” column

>     * Link to details page

Reviewer: ```suggestion
    * Link to details page
    * We will add an “exclude” list for plugins to exclude from the list. The exclusions will be stored in a file and plugin data returned will not include those excluded plugins.
```

* Error pages: Show a friendly error page if there is a site error. This should not happen often, 
  but in case it does, the user should see a friendly error page.
* Analytics tracking code: We will use Plausible without the need for cookie banners
* Deployment: work with CZI Infra’s modules for deployment to ecs cluster, we will ship our front end in a docker container
* SEO: For v0, we should give minimal attention to SEO best practices. e.g. the title should include the plugin name. 
  Each plugin should get a unique URL that includes the plugin name. plugin. the site should be readable by Google and Bing.

### Features
#### Global Search Bar
##### Technical decision
Filter on the client-side using a JS indexing library like fuse.js, lunr or minisearch.
Pros: Quick search for medium sized data and provides more complex search capabilities like fuzzy search out of the box.
Cons: Requires developer time to get familiar with the libraries and how to implement them. Libraries are fixed in what 
they can support, but they should be good enough for our needs.

##### Technical requirements

> * TBD Specifications on ordering, fuzzy search, etc. will be determined once Jeremy researches options in Solution #3.

Reviewer 2: Added! 🔥

> * TBD Specifications on ordering, fuzzy search, etc. will be determined once Jeremy researches options in Solution #3.

Reviewer: Thank you! 😃

  * PRs should ideally include E2E tests for  things like common actions, high value features, and features that depend on the backend
* Need to figure out a way to either mock the backend or setup a dev deployment for local E2E testing
* Can be used to promote staging to production:
  * Staging site can’t be promoted to production unless the E2E tests pass
* Browser Testing
  * Developers should ideally test on as many browsers they can (Chrome, Safari, and Firefox are probably fine. Edge if 
    developer has a Windows computer)
    * BrowserStack is a possible solution for testing multiple browsers
  * Would be nice to setup E2E tests for multiple browsers
    * Playwright has experimental support
    * It’s possible to test Edge since it’s a webkit browser
* Lighthouse Tests (Extra)
  * Need to setup a GitHub workflow to run Lighthouse CI for every PR
  * This will help reduce regressions in accessibility, performance, and SEO
  * Highlight improvements in aforementioned areas

#### Backend testing
* API functional testing
  * Unit test for api setup for different endpoints
* API Load testing 
  * Confirm traffic we can handle in the current setup
* Health test

>   * Setup health check for the http server

Reviewer: ```suggestion
  * Health check for the http server
```

### Release Plan

> First version of the site will be released on production by June 30, 2021.

Reviewer: ```suggestion
First version of the site was released on production on June 30, 2021. Subsequent major and minor versions to follow.
```

#### Domain name
Decision: We will use napari-hub.org with naparihub.org redirecting to napari-hub.org

> Need to work with infra team to purchase domain or use subdomain subpage of napari.org. 

Reviewer: ```suggestion
```

#### Hosting strategy
* We will use CZI hosted AWS since it will allow us more flexibility, is more future proof than Github pages and we 
  have a budget for services. We can take advantage of AWS services like Lambda and load balancing.
* We will work with CZI Infra Team to understand the components they have available for use. Terraform will be used. 
  Option to open source the Terraform files after release is still an open question on how best to do this.

#### Staging site
* We should have a separate staging site for testing
* We will likely put up a coming soon page on the production site until release

> Solutions

Reviewer: ```suggestion

### Solutions
```

 
#### API
Deploy AWS lambda (to avoid machine level maintenance) to support API where we parse pypi result with a frequency 
of 5min (configurable, fine tune later between 1min to 1 hour), and cache to aws s3. We will use the public google 

> bigquery dataset as a backup data source when the main query fails (this should also triggers an alert to our team 

Reviewer: ```suggestion
bigquery dataset as a backup data source when the main query fails (this should also trigger an alert to our team 
```

channel to indicate that the lambdas have switched to the backup approach, eng team is responsible to react to such 

> alerts and make adjustment (be it pypi outage or pypi website structure change  to address the failure and bring the

Reviewer: ```suggestion
alerts and make adjustment (be it pypi outage or pypi website structure change to address the failure and bring the
```

parsing back online). The API to be built should return a list of plugin names, in the format of 

> “[“plugin_1”, “plugin_2]”, where each individual plugin name can then be used to query qypi api

Reviewer: ```suggestion
“[“plugin_1”, “plugin_2]”, where each individual plugin name can then be used to query pypi API.
```

 
##### endpoints
* /plugins

>   * return list of plugin and versions:

Reviewer: ```suggestion
  * return list of plugin and versions
```

* /plugins/index
  * return all plugins’ metadata used in main page
* /plugins/{plugin_name}
  * return metadata for a plugin
* /plugins/{plugin_name}/version/{version}: 
  * return metadata for a specific version of plugin

#### Overall back-end technology:

> We decided to use AWS which we will use services like Lambda, S3 and API Gateway.

Reviewer: ```suggestion
We decided to use AWS which we will use services like Lambda, S3 and API Gateway. A software architecture diagram can be found [here](https://lucid.app/lucidchart/1ad7e89e-7265-4a04-b37b-135e7a57432c/view).
```

 
#### Overall front-end technology:
After building some proof of concepts and testing, we decided to use React + Next.js and TailwindCSS. Further 

> information to support this decision can be found in this document:Next.js vs SvelteKit

Reviewer: We may want to publish a summary of the findings. If we're not right now, let's remove this line.

> 

Reviewer: ```suggestion

## Publishing of plugins
* To be transparent and not require an approval process as well as share the responsibility to flagging bad plugins, we will publish activity to the hub-updates stream on the [napari Zulip](https://napari.zulipchat.com). Activity will be published for brand new plugins, new versions of plugins as well as plugin removals.

```

## Further Considerations
* Since the first version will be very minimal, consider future features that may affect the chosen technologies and engineering design.
  * Even though the first version will support a small number of plugins, we should have a plan in order to scale and 
    support potentially 1,000s of plugins in future versions.
  * We will use Plausible for analytics for the first version. In future releases, we may need more detailed analytics 
    and should provide this without the use of a cookie banner.
* Impact on other teams/projects
  * Depending on our solution to getting the plugin list data, this could be something that the napari desktop 
    application could use as well. This may help to keep consistent.
  * After starting the implementation, we learned that there is a CZI Science effort to create a unified design system 
    based on Material UI. After doing a spike to figure out how much effort and the value we would get from using this 
    framework, we decided to integrate it into our implementation. Research from the spike can be found in this 

>     document Material UI Integration

Reviewer: Same as above, if we're not going to publish the findings, let's remove this line.

* User support

> * Who is responsible for the maintenance of this site?

Reviewer: ```suggestion
For feature requests, please file [a new discussion under the ideas category](https://github.com/chanzuckerberg/napari-hub/discussions/categories/ideas).
For bug reports, please file [a new issue](https://github.com/chanzuckerberg/napari-hub/issues).
```

> ## Success Evaluation

Reviewer: ```suggestion
```

> Achieve OKR “At least 25 plugins discoverable through a plugin discovery site” by end of H1 2021

Reviewer: ```suggestion
```

### Milestones
* Milestone April 5th: Plugin Page Design Complete and Ready to Build
* Milestone May 6th: Discovery Page Design Complete and Ready to Build
* Milestone June 30th: plugin discovery site alpha/v0 release

 
### Future work
* First version will be minimal
* We will iteratively add new features and get feedback from users in a continuous cycle of future releases
 

> ## Questions

Reviewer: Please remove this entire section.
